### DISCUSSION TITLE
Question: how does Obsidian know the file path of a wikilink?

### DISCUSSION THREAD

**Discussion by zeromh** [zoni/obsidian-export#58] [Support and Q&A]
This isn't an Issue, but I'd like to understand how Obsidian.md keeps track of wikilink file paths, and how `obsidian-export` deals with that.

1. Wikilink file paths aren't included in the MD files themselves. So does Obsidian store wikilink file paths in its system files?
2. Does `obsidian-export` make use of this system information? Or does it traverse all the files in the vault and make a map of the file paths?

**Comment by zoni** [comment]
Obsidian stores all this data in the Markdown notes themselves. It may keep metadata around that as a cache for performance reasons, but all of those relations can be rebuilt from notes if the cache is deleted.

Given that Obsidian is closed-source we can't really look at how it does this without reverse-engineering, but the lookup algorithm is pretty straightforward. Here's what `obsidian-export` does:

https://github.com/zoni/obsidian-export/blob/86af6bbf3786f6259611129f7d51b572b36c947f/src/lib.rs#L722-L739

Basically, it indexes all the filenames in your vault, then finds the longest-matching `[[WikiLink]]` match, taking into account a few special rules like being case-insensitive and only have a file extension when a note isn't a Markdown file.

Obsidian itself ensures WikiLinks are unique and will prefix the directory a note is in when you have two similarly-named notes in different folders.

**Reply by zeromh** [reply]
> Obsidian stores all this data in the Markdown notes themselves.

I don't think that's true? If your link settings are set to "Shortest path when possible" (which I believe is the default), the link information in the Markdown notes doesn't contain the file location. For example, say that in the current note I make a link to a note located at `../Chemistry/crystal_structure.md`. In the current markdown file, this link looks like `[[crystal_structure]]`. No information on the location of the file. So I think Obsidian must be storing this in its system files. Am I misunderstanding something here?

> Basically, it indexes all the filenames in your vault, then finds the longest-matching [[WikiLink]] match

Just checking if I understand you... So if a file links to `[[crystal_structure]]`, `obsidian-export` will determine the location of `crystal_structure.md` because it indexes all the files in the vault. Right?
